04-materials/04-debugging.md

```diff
@@ -1,111 +1,379 @@
+
+:::{important} 👀 You should notice...
+:class: simple
+:icon: false
+
+Build-time errors appear in your terminal during `jlpm build`, while runtime errors show up in the browser console after your extension is loaded.
+:::
+
+- **Python backend errors vs TypeScript frontend errors**
+  With a full stack extension, which includes a server component, it is important to understand which side, server-side or client-side, is causing the error. Python errors will show up in the terminal from which you launched your JupyterLab instance, whereas frontend TypeScript error details will be available in the browser console. There may be times where an issue initially appears to be a frontend problem, for example, data not being displayed, but it is actually a backend problem, like an API endpoint returning the wrong data.
+
+### 📚 Reading Stack Traces
 
-### Reading Stack Traces
-- Minified vs source-mapped traces
-- JavaScript stack traces
 - Identifying the source of the error
+  When JavaScript throws an error, it gives you a stack trace, a trail showing the path the code took before it crashed. It may seem overwhelming at first, but starting from the top and looking for recognizable file names can help. While the stack trace might show you multiple function calls usually only one or two of those are in _your_ code. The others are likely in other libraries like JupyterLab itself. Finding your line in the code, you can find out where things have gone wrong.
+
+- Minified vs source-mapped traces
+  Also important, if you see names like `bundle.js:4323`, that's minified code. In development mode with source maps enabled, you'll see actual files names like `Panel.tsx:53`, a much more helpful identifier. Extensions created with the `extension-template` should have source maps enabled already.
+
+## 2️⃣ Terminal
 
-## 2. Terminal
+### 📋 Terminal Logs
 
-### Terminal Logs
 - Where to find JupyterLab server logs
+  In the terminal where you run `jupyter lab`, you'll find a great deal of information.
+
+During build time, webpack will tell you about compilation errors, missing dependencies, or TypeScript type errors. And while the messages can be verbose, you can scroll up to the first error and work through any subsequent errors.
+
+## 3️⃣ Browser Developer Tools
+
+This is where the real magic happens! Head to your browser and open the developer tools. Let's take a look at some of the most useful tabs.
+
+:::{tip} 🔧 How to open DevTools
 
-## 3. Browser Developer Tools
+- **Windows/Linux**: Press `F12` or `Ctrl+Shift+I`
+- **Mac**: Press `Cmd+Option+I`
+  :::
+
+### 💬 Console Tab
 
-### Console Tab
 - Viewing runtime errors and warnings
-- Using `console.log()`, `console.error()`, and `console.warn()`
-    - Filtering console messages
+
+Sometimes, the best tool for debugging is a simple `console.log()`. Walking through your changes and adding a logging statement to verify/inspect data, or ensure your code reaches certain logic can be quick and effective.
```

Comment:
```suggestion
Sometimes, the best tool for debugging is a simple `console.log()`, which is like a print statement for the browser console. Walking through your changes and adding a logging statement to verify/inspect data, or ensure your code reaches certain logic can be quick and effective.
```

Reply:
We could name this technique (print debugging) and include it in the glossary, too!